bench/Main.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE ImportQualifiedPost #-}
+
+module Main (main) where
+
+import Benchmark
+import Plutarch
+import Plutarch.Bool
+import Plutarch.Builtin
+import Plutarch.Integer
+import Plutarch.List qualified as List
+
+main :: IO ()
+main = do
+  benchMain benchmarks
+
+benchmarks :: [NamedBenchmark]
+benchmarks =
+  benchGroup
+    "types"
+    [ benchGroup "int" integerBench
+    , benchGroup "bool" boolBench
+    , benchGroup "builtin:list" listBench
+    ]
+
+integerBench :: [[NamedBenchmark]]
+integerBench =
+  [ -- Calling add twice
+    benchGroup
+      "add(2)"
+      $ let addInlined :: Term s PInteger -> Term s PInteger -> Term s PInteger
+            addInlined x y = x + y + 1
+            addUnhoisted :: Term s (PInteger :--> PInteger :--> PInteger)
+            addUnhoisted = plam $ \x y -> x + y + 1
+            addHoisted :: Term s (PInteger :--> PInteger :--> PInteger)
+            addHoisted = phoistAcyclic $ plam $ \x y -> x + y + 1
+         in [ bench "inlined" $ addInlined 12 32 + addInlined 5 4
+            , bench "unhoist" $ addUnhoisted # 12 # 32 + addUnhoisted # 5 # 4
+            , bench "hoisted" $ addHoisted # 12 # 32 + addHoisted # 5 # 4
+            ]
+  ]
+
+boolBench :: [[NamedBenchmark]]
+boolBench =
+  let true = pconstant @PBool True
+      false = pconstant @PBool False
+      pandNoHoist = phoistAcyclic $ plam $ \x y -> pif' # x # y # (pdelay $ pcon PFalse)
+   in [ benchGroup
+          "and"
+          [ bench "strict" $ pand' # true # false
+          , bench "lazy" $ (#&&) true false
+          , -- Calling `pand` twice.
+            bench "pand(2)" $
+              let x = pand # true # pdelay false
+               in pand # true # x
+          , bench "pand(2):unhoisted" $
+              let x = pandNoHoist # true # pdelay false
+               in pandNoHoist # true # x
+          ]
+      ]
+
+listBench :: [[NamedBenchmark]]
+listBench =
```

Comment:
Could you add the following few benches and put the report as a comment in this PR? I used these for `PIsList` optimization stuff and it'd be nice to have the report for the *current* version of the utilities.
```hs
-- `plength`
> scriptSize $ compile $ plength @PBuiltinList @PInteger
> plength # pconstant @(PBuiltinList PInteger) [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]

-- `pelem`
> scriptSize $ compile $ pelem @PBuiltinList @PInteger
> pelem # 1 # pconstant @(PBuiltinList PInteger) [5, 2, 3, 4, 7, 5, 1, 6, 2]

-- `pall`
> scriptSize $ compile $ pall @PBuiltinList @PInteger
> pall @PBuiltinList @PInteger # plam (const $ ptrace "foo" $ pconstant False) # pconstant [1, 2, 3, 4, 5, 6]

-- `plistEquals`
> scriptSize $ compile $ plistEquals @PBuiltinList @PInteger
> plistEquals @PBuiltinList @PInteger # pconstant [1, 2, 3] # pconstant [1, 2, 3]
> plistEquals @PBuiltinList @PInteger # pconstant [1, 2, 3, 4] # pconstant [1, 2, 3]
> plistEquals @PBuiltinList @PInteger # pconstant [] # pconstant [1, 2, 3]
```

Reply:
> ```haskell
> scriptSize $ compile $ plength @PBuiltinList @PInteger
> plength # pconstant @(PBuiltinList PInteger) [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
> ```

Do you really want to benchmark the lambda separately? There isn't much difference in code size between `plength` and `plength # someList`, since the applied version will differ only slightly in regards to code size (proportional to applied arguments).

In other words, we can add this benchmark group:

```haskell
      , benchGroup
          "plength"
          [ bench "lam" $ List.plength @PBuiltinList @PInteger
          , bench "app" $ List.plength # pconstant @(PBuiltinList PInteger) [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
          ]
```

Or discard the lambda entirely:

```haskell
      , bench "plength" $ List.plength # pconstant @(PBuiltinList PInteger) [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
```

Reply:
I will remove the `ptrace "foo"`, since we want to benchmark traces (if at all) separately.

Reply:
Oh yeah it's totally fine to bench the application for both script size and cpu/mem metrics, that was just a rough example.

I forgot to remove the traces in the example haha. You can remove it. It was just for checking to see short circuiting.

Reply:
The benchmark already includes script size, along with cpu/mem metrics, for the script being benchmarked. The question really is: do we want to bench _both_ applied and unapplied script? In numbers I have currently:

```
types:builtin:intlist:pelem:lam                      29873(cpu)     200(mem)    54(bytes) 
types:builtin:intlist:pelem:app                      12539807(cpu)  29554(mem)  70(bytes) 
types:builtin:intlist:pall:lam                       29873(cpu)     200(mem)    67(bytes) 
types:builtin:intlist:pall:app                       13092018(cpu)  35114(mem)  79(bytes) 
types:builtin:intlist:plistEquals:lam                1250566(cpu)   4300(mem)   167(bytes)
types:builtin:intlist:plistEquals:app:==(n=3)        25843624(cpu)  59392(mem)  180(bytes)
types:builtin:intlist:plistEquals:app:/=(n=3)        12939651(cpu)  30928(mem)  181(bytes)
types:builtin:intlist:plistEquals:app:/=(empty;n=3)  7497147(cpu)   19064(mem)  177(bytes)
```

Note that `lam` refers to unapplied script (ie., your `scriptSize $ compile $ plength @PBuiltinList @PInteger`), and `app` is the same applied on arguments (ie., `plength # pconstant @(PBuiltinList PInteger) [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]`). Did you really want both? Or can I get rid of `lam`?